Prefer ${prgnam}.SlackBuild over ${prgnam}

ErichRitz opened the pull request:
Prefer {$prgnam}.SlackBuild over ${prgnam} as name of the SlackBuild file, as old logic doesn't work with gcompris-qt.  It includes a file named gcompris-qt which slackrepo tries to use as the SlackBuild and fails.

ErichRitz committed:
Prefer ${prgnam}.SlackBuild over ${prgnam}

Prefer {$prgnam}.SlackBuild over ${prgnam} as name of the SlackBuild file, as old logic doesn't work with gcompris-qt.  It includes a file named gcompris-qt which slackrepo tries to use as the SlackBuild and fails.

aclemons commented:
Thanks for this. I still have a draft email I have been meaning to send you related to this.

I am not opposed to merging this - it seems sensible. I had just been unable to reproduce. What version of slackrepo and hints are you using?

```
# slackrepo info

slackrepo-d64ca850a64b986b8a58c2b55ff029661a01deb0

7c34928c64d1
  OS: slackware-15.0
  kernel: 5.15.94
  arch: x86_64
  nproc: 12
  total MHz: 27752
  overlayfs: yes

Configuration files
  /root/.slackreporc  [not found]
  /root/.genreprc  [not found]
  /etc/slackrepo/slackrepo_SBo.conf  [found]

Configuration options and variables
  --repo=SBo
  --lint=sb,dl,x,net,pkg,inst
  --chroot=/
  SBREPO="/var/lib/slackrepo/SBo/slackbuilds"  [not found]
  SRCREPO="/var/lib/slackrepo/SBo/source"  [not found]
  PKGREPO="/var/lib/slackrepo/SBo/packages/15.0/x86_64"  [not found]
  PKGBACKUP="/var/lib/slackrepo/SBo/backups/15.0/x86_64"  [not found]
  HINTDIR="/etc/slackrepo/SBo/hintfiles"  [not found]
  DEFAULT_HINTDIR="/etc/slackrepo/SBo/default_hintfiles/15.0"
  LOGDIR="/var/log/slackrepo/SBo"  [not found]
  DATABASE="/var/lib/slackrepo/SBo/database_SBo_x86_64.sqlite3"  [not found]
  TMP="/tmp/SBo"
  ARCH="x86_64"
  TAG="_SBo"
  PKGTYPE="txz"
  NUMJOBS="-j24 -l13"
  SUBSTITUTE=""
  USE_GENREPOS="0"

# slackrepo build gcompris-qt
Creating SlackBuild repository: /var/lib/slackrepo/SBo/slackbuilds
Creating source repository: /var/lib/slackrepo/SBo/source
Creating package repository: /var/lib/slackrepo/SBo/packages/15.0/x86_64
Creating backup repository: /var/lib/slackrepo/SBo/backups/15.0/x86_64
Creating hintfile directory: /etc/slackrepo/SBo/hintfiles
Creating database: /var/lib/slackrepo/SBo/database_SBo_x86_64.sqlite3
Populating the buildsecs table ... done.
Cloning SlackBuilds repository /var/lib/slackrepo/SBo/slackbuilds from git://git.slackbuilds.org/slackbuilds.
Cloning into '/var/lib/slackrepo/SBo/slackbuilds'...
Updating files: 100% (44143/44143), done.
Fetching origin
Switched to a new branch '15.0'
branch '15.0' set up to track 'origin/15.0'.
Finished cloning SlackBuilds repository.
Checking whether git is clean ... yes.
Indexing SlackBuilds ... done.
git repo: /var/lib/slackrepo/SBo/slackbuilds
branch:   15.0
date:     Fri Jun 23 00:49:20 GMT 2023
revision: c59da37f2f4e7609425c65ba492c542d91b5b7d8
title:    Public www update: Fri Jun 23 00:48:58 UTC 2023.

================================================================================
academic/gcompris-qt                                                    11:08:26
================================================================================

Calculating dependencies ... none.

Building academic/gcompris-qt (add version 3.3) ----------------------- 11:08:26
Testing SlackBuild files ... done.
Downloading source files ...
Verifying source files ... done.
mount: /tmp/slackrepo.pZgvYi/chrootdir/root/.Xauthority: special device /root/.Xauthority does not exist.
Running gcompris-qt.SlackBuild ...                                 eta ~11:09:??
env MAKEFLAGS='-j24 -l13' nice -n 5 bash ./gcompris-qt.SlackBuild
Built ok:  gcompris-qt-3.3-x86_64-1_SBo.txz                             11:09:26
Testing package gcompris-qt-3.3-x86_64-1_SBo.txz ... done.
Test installing gcompris-qt-3.3-x86_64-1_SBo.txz ...
gcompris-qt-3.3-x86_64-1_SBo: an educational software suite ............................................................................................................................................................................................................. [ 155M]
Unmounting chroot ... done.
:-) academic/gcompris-qt Added version 3.3 (an educational software suite) (-:

================================================================================
SUMMARY                                                                 11:09:38
================================================================================

Logfile:     /var/log/slackrepo/SBo/slackrepo_2023-06-28_11:08:04.log
Runtime:     0h 1m 34s
Added:       1
  academic/gcompris-qt/gcompris-qt-3.3-x86_64-1_SBo.txz
```

ErichRitz commented:
I'm using the latest version of slackrepo from SBo, and I build inside a chroot.  (Also note my branch 15.0-etr has minimal patches on top of upstream 15.0, none affecting gcompris-qt.)

Also I just noticed building it for the first time works, but then running update is what fails!  It order to force an update I added DOWNLOAD_x86_64 to the hintfile:

```
# slackrepo info

slackrepo-20230107

Thelio-PC
  OS: slackware-15.0
  kernel: 6.3.8-etr
  arch: x86_64
  nproc: 88
  total MHz: 204158
  overlayfs: yes

Configuration files
  /root/.slackreporc  [found]
  /root/.genreprc  [not found]
  /etc/slackrepo/slackrepo_etr.conf  [found]

Configuration options and variables
  --repo=etr
  --chroot=/
  SBREPO="/var/lib/slackrepo/etr/slackbuilds"
  SRCREPO="/var/lib/slackrepo/etr/source"
  PKGREPO="/var/lib/slackrepo/etr/packages/15.0/x86_64"
  PKGBACKUP="/var/lib/slackrepo/etr/backups/15.0/x86_64"
  HINTDIR="/var/lib/slackrepo/etr/hintfiles"
  DEFAULT_HINTDIR="/etc/slackrepo/etr/default_hintfiles/15.0"  [not found]
  LOGDIR="/var/log/slackrepo/etr"
  DATABASE="/var/lib/slackrepo/etr/database_etr_x86_64.sqlite3"
  TMP="/tmp/slackrepo/etr"
  ARCH="x86_64"
  TAG="_SBo"
  PKGTYPE="txz"
  NUMJOBS="-j176 -l89"
  SUBSTITUTE=""
  USE_GENREPOS="0"

git repo:   /var/lib/slackrepo/etr/slackbuilds
  branch:   15.0-etr
  date:     Fri Jun 23 13:14:38 -00 2023
  revision: 4240ca085c33c6791e358d1007163ff2db24f286
  title:    Merge branch 'mathics' into 15.0-etr


# cat /var/lib/slackrepo/etr/hintfiles/gcompris-qt.hint 
cat: /var/lib/slackrepo/etr/hintfiles/gcompris-qt.hint: No such file or directory
# slackrepo remove gcompris-qt
Cleaning /tmp/slackrepo/etr ... done.
git repo: /var/lib/slackrepo/etr/slackbuilds
branch:   15.0-etr
date:     Fri Jun 23 13:14:38 -00 2023
revision: 4240ca085c33c6791e358d1007163ff2db24f286
title:    Merge branch 'mathics' into 15.0-etr

================================================================================
academic/gcompris-qt                                                    13:31:00
================================================================================

Package gcompris-qt-3.3-x86_64-1_SBo.txz has been backed up and removed
Source file gcompris-qt-3.3-Linux64.sh has been backed up and removed
:-) academic/gcompris-qt Removed (-:

================================================================================
SUMMARY                                                                 13:31:00
================================================================================

Logfile:     /var/log/slackrepo/etr/slackrepo_2023-06-28_13:30:59.log
Runtime:     0h 0m 1s
Removed:     1
  academic/gcompris-qt/gcompris-qt-3.3-x86_64-1_SBo.txz

# slackrepo build gcompris-qt
Cleaning /tmp/slackrepo/etr ... done.
Checking whether git is clean ... no.
Indexing SlackBuilds ... done.
git repo: /var/lib/slackrepo/etr/slackbuilds
branch:   15.0-etr
date:     Fri Jun 23 13:14:38 -00 2023
revision: 4240ca085c33c6791e358d1007163ff2db24f286 (DIRTY)
title:    Merge branch 'mathics' into 15.0-etr
WARNING: Packages with tag _SBo are already installed
  slackrepo-20230107-noarch-1_SBo

================================================================================
academic/gcompris-qt                                                    13:31:19
================================================================================

Calculating dependencies ... none.

Building academic/gcompris-qt (add version 3.3) ----------------------- 13:31:19
Downloading source files ...
Verifying source files ... done.
Running gcompris-qt.SlackBuild ...                                  ETA 13:33:??
env MAKEFLAGS='-j176 -l89' nice -n 5 bash ./gcompris-qt.SlackBuild
Build time 407 sec, peak load 39.53, peak memory 971M, peak tmp 156M
Built ok:  gcompris-qt-3.3-x86_64-1_SBo.txz                             13:38:24
Unmounting chroot ... done.
:-) academic/gcompris-qt Added version 3.3 (an educational software suite) (-:

================================================================================
SUMMARY                                                                 13:38:31
================================================================================

Logfile:     /var/log/slackrepo/etr/slackrepo_2023-06-28_13:31:17.log
Runtime:     0h 7m 14s
Added:       1
  academic/gcompris-qt/gcompris-qt-3.3-x86_64-1_SBo.txz
Warnings:    1
  Packages with tag _SBo are already installed

# cat /var/lib/slackrepo/etr/slackbuilds/academic/gcompris-qt/gcompris-qt.info 
PRGNAM="gcompris-qt"
VERSION="3.3"
HOMEPAGE="https://gcompris.net/"
DOWNLOAD="https://gcompris.net/download/qt/linux/gcompris-qt-3.3-Linux32.sh"
MD5SUM="78c76c7f4d2086abe28a0f9a575b6ce8"
DOWNLOAD_x86_64="https://gcompris.net/download/qt/linux/gcompris-qt-3.3-Linux64.sh"
MD5SUM_x86_64="5f39645c3d824cc445e0b363dda9ced0"
REQUIRES=""
MAINTAINER="Willy Sudiarto Raharjo"
EMAIL="willysr@slackbuilds.org"
# echo DOWNLOAD_x86_64="https://gcompris.net/download/qt/linux/gcompris-qt-3.3-Linux64.sh" > /var/lib/slackrepo/etr/hintfiles/gcompris-qt.hint 
# slackrepo update gcompris-qtCleaning /tmp/slackrepo/etr ... done.
Checking whether git is clean ... no.
Indexing SlackBuilds ... done.
git repo: /var/lib/slackrepo/etr/slackbuilds
branch:   15.0-etr
date:     Fri Jun 23 13:14:38 -00 2023
revision: 4240ca085c33c6791e358d1007163ff2db24f286 (DIRTY)
title:    Merge branch 'mathics' into 15.0-etr
WARNING: Packages with tag _SBo are already installed
  slackrepo-20230107-noarch-1_SBo

================================================================================
academic/gcompris-qt                                                    13:49:34
================================================================================

Calculating dependencies ... 
Hints for academic/gcompris-qt:
  

Building academic/gcompris-qt (rebuild for hintfile changes) ---------- 13:49:35
Verifying source files ... done.
WARNING: academic/gcompris-qt: no "BUILD=" in gcompris-qt; using 1
Running gcompris-qt ...                                             ETA 13:56:??
env MAKEFLAGS='-j176 -l89' nice -n 5 bash ./gcompris-qt
academic/gcompris-qt: gcompris-qt failed (status 127)                   13:49:36
./gcompris-qt: line 3: cd: /opt/gcompris-qt/bin: No such file or directory
./gcompris-qt: line 4: ./gcompris-qt.sh: No such file or directory
See /var/log/slackrepo/etr/academic/gcompris-qt/update.log
Unmounting chroot ... done.
:-( academic/gcompris-qt FAILED )-:

================================================================================
SUMMARY                                                                 13:49:36
================================================================================

Logfile:     /var/log/slackrepo/etr/slackrepo_2023-06-28_13:49:32.log
Runtime:     0h 0m 5s
Removed:     0
Failed:      1
  academic/gcompris-qt
Warnings:    2
  Packages with tag _SBo are already installed
  academic/gcompris-qt: no "BUILD=" in gcompris-qt; using 1
```

I also just noticed an extra directory in my slackbuilds folder:
```
# ls -la /var/lib/slackrepo/etr/slackbuilds/_x86_64/
total 82816
drwxr-xr-x  2 root root     4096 Jun 28 08:31 ./
drwxr-xr-x 24 root root     4096 Jun 28 08:31 ../
-rw-r--r--  1 root root        4 Jun 12 10:07 .version
-rw-r--r--  1 root root 84783729 Jun  5 11:13 gcompris-qt-3.3-Linux64.sh
```
Based on the timestamp that was created during the build for gcompris-qt.  I removed it an reran the update, but the update still failed.

I ran "slackrepo remove gcompris-qt" and then "slackrepo build gcompris-qt" and observed the "../slackbuilds/_x86_64" directory appeared during the "Downloading source files ..." stage of the build.

aclemons commented:
Ok, the reason it behaves differently for the initial build and update is because of this code:

https://github.com/aclemons/slackrepo/blob/94fe996ff84025a395a78348481ac78e4d272f86/libexec/functions.d/parsefunctions.sh#L82-L93

and then the lookup here here based on the different flags used:

https://github.com/aclemons/slackrepo/blob/94fe996ff84025a395a78348481ac78e4d272f86/libexec/functions.d/parsefunctions.sh#L161-L175

The former yields:

```sh
sqlite> select relpath from slackbuilds where '/'||relpath||'/' glob '*/gcompris-qt/*' order by relpath asc;
academic/gcompris-qt/gcompris-qt.SlackBuild
```

while the latter:

```sh
sqlite> select distinct itemid from packages where '/'||itemid||'/' glob '*/gcompris-qt/*' order by itemid asc;
academic/gcompris-qt
```

This then has slightly different handling in the section your are changing in this PR:

https://github.com/aclemons/slackrepo/blob/94fe996ff84025a395a78348481ac78e4d272f86/libexec/functions.d/parsefunctions.sh#L202-L217

Now, I have no idea if there was a reason Dave did this check in the order it is in. If so, it was probably to do with non SBo repos or just directories of slackbuilds. Either, I'm fine with reversing the order here and if anyone was relying on it, maybe we will get an issue opened and an explanation of why it is useful.

Sorry for the delay in merging this.

aclemons reviewed (approved):
LGTM
